src/Elastic.Apm/Metrics/MetricsProvider/FreeAndTotalMemoryProvider.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using Elastic.Apm.Api;
+using Elastic.Apm.Metrics.Windows;
+
+namespace Elastic.Apm.Metrics.MetricsProvider
+{
+	/// <summary>
+	/// Returns total and free system memory.
+	/// Currently Windows & Linux -only, no macOS support at the moment.
+	/// </summary>
+	internal class FreeAndTotalMemoryProvider : IMetricsProvider
+	{
+		private const string FreeMemory = "system.memory.actual.free";
+		private const string TotalMemory = "system.memory.total";
+
+		public int ConsecutiveNumberOfFailedReads { get; set; }
+		public string NameInLogs => "total and free memory";
+
+		public IEnumerable<Sample> GetValue()
+		{
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```

Comment:
Regarding splitting the Providers into OS specific classes: With the current approach it’s easy to know where things are implemented. In case of System CPU it’s in the `SystemTotalCpuProvider`, and then you look for OS specific branches. In case we’d have multiple classes the problem is we don’t have 1-1 mapping to OSs. E.g. for System CPU there would be: 
- Windows
- Linux
- Rest of the things

In case of Total memory, only Windows, but nothing else. Then for TotalCPU we would have only xplat version. 

And with that it’ll be a mess and hard to find let’s say, “where is total CPU osx implemented?” With this, you only need to look into 1 type, it either has it, or we don’t have support for this. Plus this is again something that can be done in a subsequent PR.

I feel that we discuss in circles and this starts getting very unproductive. Therefore let’s make this quick: I don’t think that refactoring this adds value, the code will be more complicated, no benefit for the user and it takes effort. If you want me to do it I’ll do it, otherwise please accept as it is. Your call.

Reply:
I can live with it in one class for now.

Reply:
Thank you! 🙇